rust/kernel/print.rs

```diff
@@ -0,0 +1,775 @@
+        assert!(kernel_prefix[2] == b'\x00');
+
+        const SUFFIX: &[u8; LENGTH - LENGTH_PREFIX] = b"%.*s: %.*s\0";
+        [
+            kernel_prefix[0],
+            kernel_prefix[1],
+            SUFFIX[0],
+            SUFFIX[1],
+            SUFFIX[2],
+            SUFFIX[3],
+            SUFFIX[4],
+            SUFFIX[5],
+            SUFFIX[6],
+            SUFFIX[7],
+            SUFFIX[8],
+            SUFFIX[9],
+            SUFFIX[10],
+        ]
+    }
+
+    // Generate the format strings at compile-time.
+    //
+    // This avoids the compiler generating the contents on the fly in the stack.
+    //
+    // Furthermore, `static` instead of `const` is used to share the strings
+    // for all the kernel.
+    pub static EMERG: [u8; LENGTH] = generate(bindings::KERN_EMERG);
+    pub static ALERT: [u8; LENGTH] = generate(bindings::KERN_ALERT);
+    pub static CRIT: [u8; LENGTH] = generate(bindings::KERN_CRIT);
+    pub static ERR: [u8; LENGTH] = generate(bindings::KERN_ERR);
+    pub static WARNING: [u8; LENGTH] = generate(bindings::KERN_WARNING);
+    pub static NOTICE: [u8; LENGTH] = generate(bindings::KERN_NOTICE);
+    pub static INFO: [u8; LENGTH] = generate(bindings::KERN_INFO);
+    pub static DEBUG: [u8; LENGTH] = generate(bindings::KERN_DEBUG);
+
+    /// The length of the fixed format strings for the `CONT` "level".
+    pub const LENGTH_CONT: usize = 7;
+
+    /// Generates the fixed format string for the kernel's [`printk`] for
+    /// the `CONT` "level".
+    ///
+    /// [`printk`]: ../../../../include/linux/printk.h
+    const fn generate_cont() -> [u8; LENGTH_CONT] {
+        // Ensure the `KERN_CONT` macro is what we expect.
+        assert!(bindings::KERN_CONT[0] == b'\x01');
+        assert!(bindings::KERN_CONT[1] == b'c');
+        assert!(bindings::KERN_CONT[2] == b'\x00');
+
+        const SUFFIX: &[u8; LENGTH_CONT - LENGTH_PREFIX] = b"%.*s\0";
+        [
+            bindings::KERN_CONT[0],
+            bindings::KERN_CONT[1],
+            SUFFIX[0],
+            SUFFIX[1],
+            SUFFIX[2],
+            SUFFIX[3],
+            SUFFIX[4],
+        ]
+    }
+
+    // Generate the format string at compile-time for the `CONT` "level".
+    //
+    // This avoids the compiler generating the contents on the fly in the stack.
+    //
+    // Furthermore, `static` instead of `const` is used to share the string
+    // for all the kernel.
+    pub static CONT: [u8; LENGTH_CONT] = generate_cont();
+}
+
+/// Prints a message via the kernel's [`printk`].
+///
+/// Hidden because it is called from a public macro -- it should never be
+/// called by users.
+///
+/// # Safety
+///
+/// The format string must be one of the ones in [`format_strings`].
+///
+/// (This ensures that the format string matches the variadic arguments we pass
+/// to [`printk`]).
+///
+/// [`printk`]: ../../../../include/linux/printk.h
+#[doc(hidden)]
+pub unsafe fn call_printk(
```

Comment:
nit: should we return the characters written? IOW, should we convey the return of `printk` to the caller?

(Same question for the instance below.)

Reply:
Good question. Hmm... from a quick grep `= printk(`, I could only find three files. So it does not seem that common. We could nevertheless do it, or just wait until someone actually needs that.

Reply:
Returning the value seems harmless though. Is there a reason *not* to do it?

In any case, it's just a nit, feel free to leave it as is if you're too busy.

Reply:
(Taking again a look at this...)

That's fair -- I can come up with several reasons to avoid doing it, none strong, granted:
  - The `println!` etc. macros do not return anything (the kernel ones do, though).
  - To avoid dead code since nobody will use it today, i.e. YAGNI etc.; we would need more tests to do it properly, etc.
  - To avoid mistakes (`let_unit_value` lint etc.) and people starting to rely on knowing the returned size (which discourages complex formatting).
  - The use case is typically making fancy table/tree/formatting at runtime, but perhaps we could come up with something better for that.
  - The return type is not clear: `printk()` does not seem to fail, but to be on the safe side, we would need a `KernelResult`. Perhaps it is best to discuss it first with people upstream and see if we can make the guarantee explicit, so that we do not need the `KernelResult` to being with.